### ISSUE TITLE
nextEventMatchingMask should only be called from the Main Thread! (MacOS)

### ISSUE THREAD

**Issue by besser** [hybridgroup/gocv#31]
Hi, I got this error when run the 'motion-detect' demo on MacOS 10.13.1:

```bash
$ go run main.go 0
Start reading camera device: 0
2017-11-19 03:50:16.083 main[29761:4138352] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'nextEventMatchingMask should only be called from the Main Thread!'
*** First throw call stack:
(
	0   CoreFoundation                      0x00007fff4ba062fb __exceptionPreprocess + 171
	1   libobjc.A.dylib                     0x00007fff72373c76 objc_exception_throw + 48
	2   AppKit                              0x00007fff497322cf -[NSApplication(NSEvent) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 4167
	3   libopencv_highgui.3.3.dylib         0x0000000004e3bc0f cvWaitKey + 351
	4   libopencv_highgui.3.3.dylib         0x0000000004e372ca _ZN2cv9waitKeyExEi + 34
	5   libopencv_highgui.3.3.dylib         0x0000000004e37330 _ZN2cv7waitKeyEi + 34
	6   main                                0x00000000040a4e03 Window_WaitKey + 19
	7   main                                0x00000000040a026c _cgo_c1a8f8376f83_Cfunc_Window_WaitKey + 28
	8   main                                0x0000000004052970 runtime.asmcgocall + 112
)
libc++abi.dylib: terminating with uncaught exception of type NSException
SIGABRT: abort
PC=0x7fff730b1fce m=3 sigcode=0
signal arrived during cgo execution

goroutine 1 [syscall, locked to thread]:
runtime.cgocall(0x40a0250, 0xc42003ae30, 0x0)
	/usr/local/Cellar/go/1.9.2/libexec/src/runtime/cgocall.go:132 +0xe4 fp=0xc42003adf0 sp=0xc42003adb0 pc=0x4004334
gocv.io/x/gocv._Cfunc_Window_WaitKey(0x1, 0x0)
	gocv.io/x/gocv/_obj/_cgo_gotypes.go:1653 +0x4d fp=0xc42003ae30 sp=0xc42003adf0 pc=0x409d07d
gocv.io/x/gocv.WaitKey(0x1, 0x9f0ca90)
	/Users/besser/go/src/gocv.io/x/gocv/highgui.go:128 +0x2a fp=0xc42003ae50 sp=0xc42003ae30 pc=0x409d65a
main.main()
	/Users/besser/go/src/gocv.io/x/gocv/cmd/motion-detect/main.go:100 +0x5d5 fp=0xc42003af80 sp=0xc42003ae50 pc=0x409f7b5
runtime.main()
	/usr/local/Cellar/go/1.9.2/libexec/src/runtime/proc.go:195 +0x226 fp=0xc42003afe0 sp=0xc42003af80 pc=0x402b986
runtime.goexit()
	/usr/local/Cellar/go/1.9.2/libexec/src/runtime/asm_amd64.s:2337 +0x1 fp=0xc42003afe8 sp=0xc42003afe0 pc=0x40537d1

rax    0x0
rbx    0x70000c346000
rcx    0x70000c344428
rdx    0x0
rdi    0xc03
rsi    0x6
rbp    0x70000c344460
rsp    0x70000c344428
r8     0x70000c3442f0
r9     0x70000c3444c0
r10    0x0
r11    0x206
r12    0xc03
r13    0x30
r14    0x6
r15    0x2d
rip    0x7fff730b1fce
rflags 0x206
cs     0x7
fs     0x0
gs     0x0
exit status 2
```

**Comment by deadprogram** [comment]
Hi, @besser thanks for the report.

Can you try adding the following line as the first line in your `main()` as follows:

```go
runtime.LockOSThread()
```

You will also need to add the `"runtime"` package to your list of imports.

Please let me know of that has any effect. I do not have an install of High Sierra currently for testing, so your feedback is appreciated.

**Comment by besser** [comment]
Yes, it's worked!

But it was necessary to comment on line 46 (49 after added the code indicated by you) because appeared only a white blank in full screen:

```go
window.SetWindowProperty(gocv.WindowPropertyFullscreen, gocv.WindowFullscreen)
```

**Comment by deadprogram** [comment]
Both of those changes, correct?

**Comment by deadprogram** [comment]
That is very interesting, as the call to `runtime.LockOSThread()` is not needed on earlier verrsion of OS X...

**Comment by deadprogram** [comment]
HI @besser can you please try the latest `dev` branch without your last changes? I already removed the `window.SetWindowProperty()` line.

I am just curious to see if the threading is still a problem with the latest changes made to the BackgroundSelectorMOG2. Perhaps you can check?

Thanks again!

**Comment by besser** [comment]
Hi @deadprogram, I changed to branch `dev` and I got same error... :(

```bash
╭─besser@Buran ~/go/src/gocv.io/x/gocv/cmd/motion-detect  ‹dev›
╰─$ go run main.go 0
Start reading camera device: 0
2017-11-23 01:09:18.371 main[18241:439981] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'nextEventMatchingMask should only be called from the Main Thread!'
*** First throw call stack:
(
	0   CoreFoundation                      0x00007fff52cb02fb __exceptionPreprocess + 171
	1   libobjc.A.dylib                     0x00007fff7961dc76 objc_exception_throw + 48
	2   AppKit                              0x00007fff509dc2cf -[NSApplication(NSEvent) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 4167
	3   libopencv_highgui.3.3.dylib         0x000000000788cc0f cvWaitKey + 351
	4   libopencv_highgui.3.3.dylib         0x00000000078882ca _ZN2cv9waitKeyExEi + 34
	5   libopencv_highgui.3.3.dylib         0x0000000007888330 _ZN2cv7waitKeyEi + 34
	6   main                                0x00000000040a6a43 Window_WaitKey + 19
	7   main                                0x000000000409fdec _cgo_2fde5d5c1e58_Cfunc_Window_WaitKey + 28
	8   main                                0x0000000004052480 runtime.asmcgocall + 112
)
libc++abi.dylib: terminating with uncaught exception of type NSException
SIGABRT: abort
PC=0x7fff7a35bfce m=4 sigcode=0
signal arrived during cgo execution

goroutine 1 [syscall, locked to thread]:
runtime.cgocall(0x409fdd0, 0xc42003ae30, 0x0)
	/usr/local/Cellar/go/1.9.2/libexec/src/runtime/cgocall.go:132 +0xe4 fp=0xc42003adf0 sp=0xc42003adb0 pc=0x4003e44
gocv.io/x/gocv._Cfunc_Window_WaitKey(0x1, 0x0)
	gocv.io/x/gocv/_obj/_cgo_gotypes.go:2010 +0x4d fp=0xc42003ae30 sp=0xc42003adf0 pc=0x409ca8d
gocv.io/x/gocv.WaitKey(0x1, 0x975b480)
	/Users/besser/go/src/gocv.io/x/gocv/highgui.go:128 +0x2a fp=0xc42003ae50 sp=0xc42003ae30 pc=0x409cfaa
main.main()
	/Users/besser/go/src/gocv.io/x/gocv/cmd/motion-detect/main.go:99 +0x5a8 fp=0xc42003af80 sp=0xc42003ae50 pc=0x409f058
runtime.main()
	/usr/local/Cellar/go/1.9.2/libexec/src/runtime/proc.go:195 +0x226 fp=0xc42003afe0 sp=0xc42003af80 pc=0x402b496
runtime.goexit()
	/usr/local/Cellar/go/1.9.2/libexec/src/runtime/asm_amd64.s:2337 +0x1 fp=0xc42003afe8 sp=0xc42003afe0 pc=0x40532e1

rax    0x0
rbx    0x700009565000
rcx    0x700009563428
rdx    0x0
rdi    0xd03
rsi    0x6
rbp    0x700009563460
rsp    0x700009563428
r8     0x7000095632f0
r9     0x7000095634c0
r10    0x0
r11    0x206
r12    0xd03
r13    0x30
r14    0x6
r15    0x2d
rip    0x7fff7a35bfce
rflags 0x206
cs     0x7
fs     0x0
gs     0x0
exit status 2
```
... and with the call `runtime.LockOSThread()` in first line it worked well.

Thanks

**Comment by deadprogram** [comment]
@besser you are a great help here. One last thing I will ask, please. I have a new branch `feature/window-lock-thread` that should correct this without any further changes. Can you please pull this branch, and then run motion-detect without any changes?

Note that `WaitKey()` has been deprecated in favor of `Window.WaitKey()`: https://github.com/hybridgroup/gocv/blob/feature/window-lock-thread/cmd/motion-detect/main.go#L99

Thank you!

**Comment by besser** [comment]
Hey @deadprogram, great news! I ran montion-detect from branch `feature/window-lock-thread` without problems! :)

Thanks again!

**Comment by deadprogram** [comment]
This was released with v0.6.0 yesterday so now closing. Thanks again for your help @besser
